### PR TITLE
Implement a basic dry-run option for the downloader

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

> **Note**
> While the implementation is not complete, it has already served me well, so I didn't want to hold it back.

### CODE REVIEW

File: cli/src/main/kotlin/commands/DownloaderCommand.kt [hunk 1]
```diff
@@ -346,11 +346,7 @@ class DownloaderCommand : OrtCommand(
         } catch (e: DownloadException) {
             e.showStackTrace()
 
-            val failureMessage = "Could not download '${pkg.id.toCoordinates()}': " +
-                    e.collectMessages()
-            failureMessages += failureMessage
-
-            logger.error { failureMessage }
+            failureMessages += e.collectMessages()
```
Comment (mnonnenmacher) [inline cli/src/main/kotlin/commands/DownloaderCommand.kt:358]: Commit message: "as the happen" -> "when they are happening"?

File: downloader/src/main/kotlin/Downloader.kt [hunk 2]
```diff
@@ -66,7 +68,7 @@ class Downloader(private val config: DownloaderConfiguration) {
      * Download the source code of the [package][pkg] to the [outputDirectory]. A [Provenance] is returned on success or
      * a [DownloadException] is thrown in case of failure.
      */
-    fun download(pkg: Package, outputDirectory: File): Provenance {
+    fun download(pkg: Package, outputDirectory: File, dryRun: Boolean = false): Provenance {
```
Comment (mnonnenmacher) [inline downloader/src/main/kotlin/Downloader.kt:71]: Please explain the `dryRun` argument in the function docs.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: utils/ort/src/main/kotlin/OkHttpClientHelper.kt [hunk 3]
```diff
@@ -217,6 +217,14 @@ fun OkHttpClient.download(url: String, acceptEncoding: String? = null): Result<P
         response to body
     }
 
+/**
+ * Execute a HEAD-request against [url] to ping for its existence.
+ */
+fun OkHttpClient.ping(url: String): Response =
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/ort/src/main/kotlin/OkHttpClientHelper.kt:223]: "head" is not a verb, so I preferred "ping" here, even if the meaning is slightly different. I could maybe rename to `exists` (like `File` has), if you prefer.

[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply utils/ort/src/main/kotlin/OkHttpClientHelper.kt:223]: I would still prefer to stick with HTTP terminology over using a verb out of principle.

[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/ort/src/main/kotlin/OkHttpClientHelper.kt:223]: But isn't the fact that a `HEAD` request is used just an implementation detail that shouldn't necessarily be exposed via the function name?

[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/ort/src/main/kotlin/OkHttpClientHelper.kt:223]: I mean, we also have `download()` which internally uses a `GET` request.

File: cli/src/main/kotlin/commands/DownloaderCommand.kt (hunk 1 above)
Comment (sschuberth) [reply cli/src/main/kotlin/commands/DownloaderCommand.kt:358]: Corrected to say "as they happen".

File: downloader/src/main/kotlin/Downloader.kt (hunk 2 above)
Comment (sschuberth) [reply downloader/src/main/kotlin/Downloader.kt:71]: Done for all public functions.

File: downloader/src/main/kotlin/Downloader.kt [hunk 4]
```diff
@@ -303,6 +329,18 @@ class Downloader(private val config: DownloaderConfiguration) {
...
+        if (dryRun) {
+            if (sourceArchive.isFile) return ArtifactProvenance(sourceArtifact)
+
+            val response = okHttpClient.ping(sourceArtifact.url)
+
+            return if (response.code == HttpURLConnection.HTTP_OK) {
+                ArtifactProvenance(sourceArtifact)
+            } else {
+                UnknownProvenance
+            }
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply downloader/src/main/kotlin/Downloader.kt:341]: Good point, I've aligned on throwing a `DownloadException`.

File: utils/ort/src/main/kotlin/OkHttpClientHelper.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply utils/ort/src/main/kotlin/OkHttpClientHelper.kt:223]: Yes, I think I would prefer other names for those functions as well, but that's not related to this PR.

**Pull request merged**